be/src/storage/rowset/segment_v2/segment_writer.cpp

```diff
@@ -263,13 +316,43 @@ Status SegmentWriter::append_chunk(const vectorized::Chunk& chunk) {
 
     for (size_t i = 0; i < chunk.num_rows(); i++) {
         // At the begin of one block, so add a short key index entry
-        if ((_row_count % _opts.num_rows_per_block) == 0) {
+        if ((_num_rows_written % _opts.num_rows_per_block) == 0) {
             size_t keys = _tablet_schema->num_short_key_columns();
             vectorized::SeekTuple tuple(*chunk.schema(), chunk.get(i).datums());
             std::string encoded_key = tuple.short_key_encode(keys, 0);
             RETURN_IF_ERROR(_index_builder->add_item(encoded_key));
         }
-        ++_row_count;
+        ++_num_rows_written;
+    }
+    return Status::OK();
+}
+
+Status SegmentWriter::append_chunk(const vectorized::Chunk& chunk, const std::vector<uint32_t>& column_indexes,
+                                   bool is_key) {
+    DCHECK_EQ(chunk.num_columns(), column_indexes.size());
+    for (size_t i = 0; i < column_indexes.size(); ++i) {
+        uint32_t column_index = column_indexes[i];
+        if (column_index >= _column_writers.size()) {
+            return Status::InternalError("column index error");
+        }
+        const vectorized::Column* col = chunk.get_column_by_index(i).get();
+        RETURN_IF_ERROR(_column_writers[column_index]->append(*col));
+    }
+
+    size_t chunk_num_rows = chunk.num_rows();
+    if (is_key) {
+        for (size_t i = 0; i < chunk_num_rows; i++) {
+            // At the begin of one block, so add a short key index entry
+            if ((_num_rows_written % _opts.num_rows_per_block) == 0) {
+                size_t keys = _tablet_schema->num_short_key_columns();
+                vectorized::SeekTuple tuple(*chunk.schema(), chunk.get(i).datums());
+                std::string encoded_key = tuple.short_key_encode(keys, 0);
+                RETURN_IF_ERROR(_index_builder->add_item(encoded_key));
+            }
+            ++_num_rows_written;
```

Comment:
the increment of _num_rows_written can be done outside the is_key branch, with else together

Reply:
Write short key requires `++ _num_rows_written`